Fix illumos build regressions

am11 opened the pull request:
* gcc looks for fallthrough or break in morph (we could move the break out of condition, but there are a few gotos to that case which I am not sure might have sideeffect, so i kept it simple).
* SHM_ANON is not available in illumos/Solaris libc (POSIX does not require it), so manually generate name using timespec for shm_open() call.
* in line with previous fixes made for statfs, SunOS-like OS has deprecated statfs in favor of statvfs. The f_type does not return magic number as other Unices. The more reliable way is to get the stringly name from f_basetype and then resolve the enum value. Assuming performance is not a concern for DriveInfo, kept it simple.

am11 commented:
cc @janvorli,

```sh
docker run -e ROOTFS_DIR=/crossrootfs/x64 -v $(pwd):/runtime \
    mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-18.04-cross-illumos-20210714135645-f13d79e \
    /runtime/build.sh -arch x64 -os illumos -cross -gcc
```
(entire build) continue to succeed with this patch.

janvorli commented on src/native/corehost/bundle/extractor.cpp:
```diff
@@ -7,6 +7,10 @@
 #include "pal.h"
 #include "utils.h"
 
+#ifdef __sun
```
Why do we need `alloca.h` just for __sun?

am11 replied:
alloca() is defined in that header. we did the same thing elsewhere, e.g. https://github.com/dotnet/runtime/blob/d05950ba1ad440855f9851857bbad6f9d1a34321/src/coreclr/gc/env/gcenv.base.h#L14-L16

am11 commented on src/coreclr/minipal/Unix/doublemapping.cpp:
```diff
@@ -41,6 +41,32 @@ bool VMToOSInterface::CreateDoubleMemoryMapper(void** pHandle, size_t *pMaxExecu
+    char* fill;
+    int fd;
+
+    while (true)
```
Changed it to append getpid() instead of timespec. Also shm_unlink is moved before shm_open.

janvorli commented:
The changes look good, but I'd like some libraries folks to review the related part of the change. @stephentoub could you please take a look or add someone else to review this?

am11 committed:
Fix illumos build regressions

am11 commented:
Test failure on Red Hat Linux is unrelated (#55995).

am11 commented:
@stephentoub, @adamsitnik. the library change for official platforms has no diffs in IL or PAL binaries. I have split a partial class and updated the native implementation using existing macros, which indicates if system has a legacy `struct statfs` (Solaris deprecated it during the last decade) and uses `stat statvfs` instead which provides the filesystem name (string) instead of its ID (number). Please take a look. :)

stephentoub commented:
Sorry, I missed the notification when I was tagged. Looking now...

stephentoub commented on src/libraries/Native/Unix/System.Native/pal_io.c:
```diff
@@ -1406,10 +1410,28 @@ int64_t SystemNative_GetFileSystemType(intptr_t fd)
 #else
-    #error "Platform doesn't support fstatfs"
+    (void)fd; // unused
+    return -1;
```
The purpose of the #error here was to know if we were ever building against something that lacked the function we were depending on.  Don't we now silently lose that validation?

am11 replied:
We surely need that build-time validation. The error is actually moved a few lines below and its condition is changed from `#if !HAVE_STATFS_VFS && !HAVE_STATFS_MOUNT` to have ` && HAVE_NON_LEGACY_STATFS`.

stephentoub replied:
That's not the same thing, though.  Let's say we started targeting a new platform that had statvfs but not statfs, yet we were using the code file that invoked SystemNative_GetFileSystemType.  We wouldn't get a compile time error, but we would silently get wrong behavior at run time.

am11 replied:
Ah, you are right. I was not thinking about association with which partial we will use by default. I have limit the `return -1` case to SunOS platforms.

adamsitnik replied:
> Would it be better to not add the additional export and instead just do a translation here from file system name to expected ID? Or would that result in duplicating the entire enum we have on the managed side?

I would also prefer to just add a simple mapping to `SystemNative_GetFileSystemType`. We don't need to map all values, we care only about four of them:

https://github.com/dotnet/runtime/blob/28ce2cc5e487f504c8fc91515c41a5fb0e86552c/src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Unix.cs#L368-L371

am11 replied:
If we only ever care about four members, should we remove rest from `enum UnixFileSystemTypes`?

am11 replied:
Ok, I have added the conditions for existing FS types and deleted the additional export. If/when we decide to cleanup `enum UnixFileSystemTypes`, we can cleanup these conditions as well.

am11 committed:
Add back build validation in GetFileSystemType

stephentoub commented on src/libraries/Native/Unix/System.Native/pal_io.c:
```diff
@@ -1405,11 +1409,31 @@ int64_t SystemNative_GetFileSystemType(intptr_t fd)
     // which got deprecated in macOS 10.6, in favor of statfs
     while ((statfsRes = fstatfs(ToFileDescriptor(fd), &statfsArgs)) == -1 && errno == EINTR) ;
     return statfsRes == -1 ? (int64_t)-1 : (int64_t)statfsArgs.f_type;
+#elif defined(TARGET_SUNOS // we use statvfs (see SystemNative_GetFileSystemTypeAsString) on SunOS
```
```suggestion
#elif defined(TARGET_SUNOS) // we use statvfs (see SystemNative_GetFileSystemTypeAsString) on SunOS
```

stephentoub replied:
Would it be better to not add the additional export and instead just do a translation here from file system name to expected ID?

am11 replied:
It would result in duplicating the enum entries that we have on the managed side.

The `f_type` in `struct statfs` is not available with the macros we have defined. I read the code and it seems like even if we change the macro definitions for system header to reveal this field, we will still get zero: https://github.com/illumos/illumos-gate/blob/9ecd05bdc59e4a1091c51ce68cce2028d5ba6fd1/usr/src/ucbhead/sys/vfs.h#L241
cc @jclulow, @sjorge in case I've missed anything :)

stephentoub commented on src/libraries/Common/src/Interop/Unix/System.Native/Interop.UnixFileSystemTypes.statvfs.cs:
```diff
@@ -0,0 +1,18 @@
+        private static extern string GetFileSystemTypeAsString(SafeFileHandle fd);
+
+        internal static bool TryGetFileSystemType(SafeFileHandle fd, out UnixFileSystemTypes fileSystemType) =>
+            Enum.TryParse<UnixFileSystemTypes>(GetFileSystemTypeAsString(fd), out fileSystemType);
```
Are the names we have as the enum names all correct for this to work?

am11 replied:
I tested it in OpenIndiana 2021.04 VM which has the following setup:

```sh
$ df -h
Filesystem             Size   Used  Available Capacity  Mounted on
rpool/ROOT/openindiana-2021:07:18
                     47.95G  5.03G     26.76G    16%    /
/devices                  0      0          0     0%    /devices
/dev                      0      0          0     0%    /dev
ctfs                      0      0          0     0%    /system/contract
proc                      0      0          0     0%    /proc
mnttab                    0      0          0     0%    /etc/mnttab
swap                  3.58G   864K      3.58G     1%    /etc/svc/volatile
objfs                     0      0          0     0%    /system/object
bootfs                    0      0          0     0%    /system/boot
sharefs                   0      0          0     0%    /etc/dfs/sharetab
/usr/lib/libc/libc_hwcap1.so.1
                     31.79G  5.03G     26.76G    16%    /lib/libc.so.1
fd                        0      0          0     0%    /dev/fd
rpool/ROOT/openindiana-2021:07:18/var
                     47.95G  2.19G     26.76G     8%    /var
swap                  3.58G     4K      3.58G     1%    /tmp
swap                  3.58G    48K      3.58G     1%    /var/run
rpool/export         47.95G 24.50K     26.76G     1%    /export
rpool                47.95G    33K     26.76G     1%    /rpool
rpool/export/home    47.95G 24.50K     26.76G     1%    /export/home
rpool/export/home/vagrant
                     47.95G  1.08G     26.76G     4%    /export/home/vagrant
vagrant              931.55G 213.34G    718.21G    23%    /vagrant
projects             931.55G 213.34G    718.21G    23%    /projects
```

with the following C program:

```c
#include <stdio.h>
#include <errno.h>
#include <fcntl.h>
#include <inttypes.h>
#include <sys/vfs.h>

char*
// int64_t
SystemNative_GetFileSystemType(intptr_t fd)
{
    int statfsRes;
    // struct statfs statfsArgs;
    struct statvfs statfsArgs;
    // while ((statfsRes = fstatfs(fd, &statfsArgs)) == -1 && errno == EINTR) ;
    while ((statfsRes = fstatvfs(fd, &statfsArgs)) == -1 && errno == EINTR) ;
    // return statfsRes == -1 ? (int64_t)-1 : (int64_t)statfsArgs.f_type;
    return statfsRes == -1 ? "-1" : statfsArgs.f_basetype;
}

void PrintStats(const char* name) {
    int fd = open(name, O_RDONLY);

    printf("File: %s, BaseType: %s\n",
        name,
        SystemNative_GetFileSystemType((intptr_t)fd));
}

int main() {
    PrintStats("/proc");
    PrintStats("/");
    PrintStats("/tmp");
    PrintStats("/system/contract");
    PrintStats("/devices");
    PrintStats("/dev");
    PrintStats("/etc/mntab");
    PrintStats("/etc/svc/volatile");
    PrintStats("/system/object");
    PrintStats("/system/boot");
    PrintStats("/etc/dfs/sharetab");
    PrintStats("/lib/libc.so.1");
    PrintStats("/dev/fd");
    PrintStats("/var");
    PrintStats("/var/run");
    PrintStats("/rpool");
    PrintStats("/export");
    PrintStats("/export/home");
    PrintStats("/export/home/vagrant");
    PrintStats("/vagrant");

    return 0;
}
```

output of the program suggests that with the exception of procfs, all base types are reported correctly and we have them in `enum UnixFileSystemTypes`:

```
File: /proc, BaseType: 
File: /, BaseType: zfs
File: /tmp, BaseType: tmpfs
File: /system/contract, BaseType: ctfs
File: /devices, BaseType: devfs
File: /dev, BaseType: dev
File: /etc/mntab, BaseType: -1
File: /etc/svc/volatile, BaseType: tmpfs
File: /system/object, BaseType: objfs
File: /system/boot, BaseType: bootfs
File: /etc/dfs/sharetab, BaseType: sharefs
File: /lib/libc.so.1, BaseType: zfs
File: /dev/fd, BaseType: fd
File: /var, BaseType: zfs
File: /var/run, BaseType: tmpfs
File: /rpool, BaseType: zfs
File: /export, BaseType: zfs
File: /export/home, BaseType: zfs
File: /export/home/vagrant, BaseType: zfs
File: /vagrant, BaseType: sffs
```

Perhaps @jclulow might know the reason why `statvfs.f_basetype` is empty in procfs case.

am11 replied:
Good observation @jclulow, and sorry for the nuisance. I actually have `strdup()` around `statfsArgs.f_basetype` in the actual PR change, but missed it in the test example. `/proc` indeed returns `procfs`. However, `/etc/mntab` case is empty (which seems intentional?).

I couldn't find magic numbers for:

```
ctfs
dev
objfs
sffs
sharefs
```

in Linux or FreeBSD sources, so I haven't added them.  At usage side, the library code will consider them "unknown filesystems". If you know where could I find magic number for these, please let me know. Otherwise, I think "unknown" is file and as these filesystems are not too critical in the current usage context (it is mostly for completeness).

am11 replied:
> output of the program suggests that with the exception of procfs, all base types are reported correctly and we have them in `enum UnixFileSystemTypes`:

[self correction] this was wrong, as we do not have these:

```
ctfs
dev
objfs
sffs
sharefs
```

I added some of these from SmartOS in #34867, but seems like OpenIndiana has more variety.

am11 replied:
Ok, I think these were dropped during the refactoring. I have found some missing magic numbers and set rest to tmpfs magic as before. For Linux/macOS/FreeBSD, it will use magic and for SunOS-derivates, it will match the names.

am11 committed:
Update src/libraries/Native/Unix/System.Native/pal_io.c

Co-authored-by: Stephen Toub <stoub@microsoft.com>

am11 committed:
Add back some illumos/solaris filesystems

am11 committed:
Sync DriveType and UnixFileSystemTypes

am11 commented:
Windows x86 failure is unrelated (and fixed in `main` branch: #56649).

adamsitnik commented on src/libraries/Native/Unix/System.Native/pal_io.c:
```diff
@@ -1405,11 +1409,31 @@ int64_t SystemNative_GetFileSystemType(intptr_t fd)
+    int statfsRes;
+    struct statvfs statfsArgs;
+    while ((statfsRes = fstatvfs(ToFileDescriptor(fd), &statfsArgs)) == -1 && errno == EINTR) ;
+    return statfsRes == -1 ? NULL : strdup(statfsArgs.f_basetype);
```
`strdup` allocates memory that needs to be freed. How are we going to free this memory?

am11 replied:
This memory doesn't need to be freed explicitly in our code, it will be freed by the GC when managed method goes out of scope. Also, have a look at other `strdup` usages https://github.com/dotnet/runtime/search?q=strdup+path:src/libraries/native.

stephentoub replied:
> it will be freed by the GC when managed method goes out of scope

Not quite.  It'll be freed by the interop layer as part of creating a string from the returned pointer.

am11 committed:
Add FS type mapping for statvfs case

am11 committed:
Merge dotnet/main into fix/illumos-build

am11 committed:
Revert now-unchanged csproj lineending

adamsitnik commented on src/libraries/Native/Unix/System.Native/pal_io.c:
```diff
@@ -1405,8 +1409,138 @@ int64_t SystemNative_GetFileSystemType(intptr_t fd)
+    else if (strcmp(statfsArgs.f_basetype, "xia") == 0) result = 0x012FD16D;
+    else if (strcmp(statfsArgs.f_basetype, "udev") == 0) result = 0x01021994;
+    else if (strcmp(statfsArgs.f_basetype, "zfs") == 0) result = 0x2FC12FC1;
+
```
it would be nice to assert the result here to detect missing mappings (might happen in a few years from now):

```suggestion

    assert(result != -1);
```

Ideally we would include `f_basetype` in the error msg, but I am not sure if the following code is going to compile:

```c
    assert_msg(result != -1, "Unknown file system type", statfsArgs.f_basetype);
```

am11 replied:
Unfortunately, `assert_msg` only accepts `int` as third argument.

It would be good to add have string variant of that macro (or change existing `assert_*`'s to convert `int`s to string and accept only string). Maybe at the same time, we should also move `assert_*` and `c_static_assert_*` macros in a single header (rather than split them in `pal_utilities.c` and `pal_compiler.c`).

adamsitnik reviewed (approved):
System.IO part LGTM.

@am11 big thanks for another great contribution! And kudos for adding extra file system types to our existing enumeration!

am11 committed:
Update src/libraries/Native/Unix/System.Native/pal_io.c

Co-authored-by: Adam Sitnik <adam.sitnik@gmail.com>
